sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ExecutionStateSampler.java

```diff
@@ -333,31 +335,41 @@ private void takeSample(long currentTimeMillis, long millisSinceLastSample) {
         transitionsAtLastSample = transitionsAtThisSample;
       } else {
         long lullTimeMs = currentTimeMillis - lastTransitionTime.get();
-        Thread thread = trackedThread.get();
         if (lullTimeMs > MAX_LULL_TIME_MS) {
-          if (thread == null) {
-            LOG.warn(
-                String.format(
-                    "Operation ongoing in bundle %s for at least %s without outputting or completing (stack trace unable to be generated).",
-                    processBundleId.get(),
-                    DURATION_FORMATTER.print(Duration.millis(lullTimeMs).toPeriod())));
-          } else if (currentExecutionState == null) {
-            LOG.warn(
-                String.format(
-                    "Operation ongoing in bundle %s for at least %s without outputting or completing:%n  at %s",
-                    processBundleId.get(),
-                    DURATION_FORMATTER.print(Duration.millis(lullTimeMs).toPeriod()),
-                    Joiner.on("\n  at ").join(thread.getStackTrace())));
-          } else {
-            LOG.warn(
-                String.format(
-                    "Operation ongoing in bundle %s for PTransform{id=%s, name=%s, state=%s} for at least %s without outputting or completing:%n  at %s",
-                    processBundleId.get(),
-                    currentExecutionState.ptransformId,
-                    currentExecutionState.ptransformUniqueName,
-                    currentExecutionState.stateName,
-                    DURATION_FORMATTER.print(Duration.millis(lullTimeMs).toPeriod()),
-                    Joiner.on("\n  at ").join(thread.getStackTrace())));
+          if (lullTimeMs < lastLullReport // This must be a new report.
+              || lullTimeMs > 1.2 * lastLullReport // Exponential backoff.
```

Comment:
do we want to bother with the exponential? Seems more complicated than necessary and just adds 3 more logs which don't seem particularly valuable and perhaps spammier than desired:
5 min, 6 min, 7.2 min, 8.64, 10, 15, 20

Reply:
I think it's still nicer to log a bit frequently first.